### ISSUE TITLE
[Performance] SmolVLA in-place DynamicCache crop prevents torch.compile and limits inference to 4.8 Hz

### ISSUE THREAD

**Issue by NotNANtoN** [huggingface/lerobot#4633]
### Description

In `SmolVLAPolicy.denoise_step` (`lerobot/policies/smolvla/modeling_smolvla.py` and `smolvlm_with_expert.py`), `past_key_values` uses Hugging Face's `DynamicCache`.

During each step of the Euler flow-matching integration (`num_steps=10`), the action suffix key/value tokens are appended in-place via `past_key_values.update()`, and then truncated back to the prefix length via `past_key_values.crop(prefix_len)`:

```python
outputs_embeds, _ = self.vlm_with_expert.forward(
    attention_mask=full_att_2d_masks,
    position_ids=position_ids,
    past_key_values=past_key_values,
    inputs_embeds=[None, suffix_embs],
    use_cache=self.config.use_cache,
)
if past_key_values is not None:
    # Self-attention layers append suffix K/V in place; restore the prefix for the next step.
    past_key_values.crop(prefix_len)
```

### Problems Caused:
1. **Dynamic Reallocations & Crop Overhead**: Calling `update()` followed by `crop(prefix_len)` 10 times per inference chunk incurs significant Python and CUDA memory allocation overhead.
2. **Incompatible with `torch.compile`**: Because the cache tensors are mutated in-place and dynamically resized every iteration, `torch.compile` catnot capture CUDA graphs or optimize kernels across iterations (resulting in dynamic shape guard re-evaluations and slower execution).

### Proposed Solution

Make the prefix key/value cache **static and immutable**:
1. After computing the prefix KV cache once in `sample_actions()`, freeze it into a tuple of static tensors:
   ```python
   past_key_values = tuple((layer.keys, layer.values) for layer in past_key_values.layers)
   ```
2. In `forward_attn_layer` and `forward_cross_attn_layer`, when `past_key_values` is a tuple, concatenate the prefix functionally without mutating the stored prefix buffers:
   ```python
   prefix_k, prefix_v = past_key_values[layer_idx]
   key_states = torch.cat([prefix_k, key_states.transpose(1, 2)], dim=2).transpose(1, 2)
   value_states = torch.cat([prefix_v, value_states.transpose(1, 2)], dim=2).transpose(1, 2)
   ```
3. Skip `past_key_values.crop(prefix_len)` when using static tuple KV.
4. Add an optional `compile_denoise: bool = False` flag in `SmolVLAConfig` to allow compiling the denoise step with `torch.compile(mode="max-autotune")`.

### Benchmark (RTX 4090, `lerobot/smolvla_base`):
- **Eager baseline**: `207.28 ms` (4.8 Hz)
- **Static KV + `max-autotune`**: **`49.43 ms` (20.2 Hz)** — **4.19x speedup** (saving 157.85 ms per prediction)
- **Bitwise parity**: Cosine Similarity = **`0.99999`**

We have tested these changes and have a clean PR ready to link!

**Comment by sattyamjjain** [comment]
Question on the compile path, since the prefix KV length depends on the instruction's token count: does compile_denoise pad the language prefix to a fixed length, or will each new instruction length trigger a recompile (and a first-call stall inside a running control loop)? I have a Jetson Orin Nano Super 8GB on JetPack 6.x and can benchmark this branch there, eager vs compile_denoise, with timings per instruction length, if that is useful. Also: how long is the max-autotune warm-up on the 4090 before the first 49 ms prediction?

**Comment by NotNANtoN** [comment]
Great questions! Just benchmarked this directly on an RTX 4090 with `lerobot/smolvla_base`:

1. **Instruction Length & Recompilation**:
   - In the official `lerobot/smolvla_base` checkpoint, `pad_language_to` is set to `"max_length"` (`tokenizer_max_length = 48`), so the prefix token length is static at 48 tokens across tasks by default.
   - Even if `pad_language_to="longest"` is used (variable-length tokens), PyTorch Inductor treats the sequence dimension as dynamic. Testing a switch from a 2-token prompt to a 10-token prompt executed in **2.4 ms** with no recompile or stall.

2. **Warm-up Time**:
   - The first call with `torch.compile(mode="max-autotune")` takes **~7–8 seconds** on an RTX 4090 to benchmark candidate Triton kernels.
   - Once warmed up, end-to-end action prediction drops from **188.9 ms (5.3 Hz)** down to **46.1 ms (21.7 Hz)** (~4.1x speedup).

3. **Jetson Orin Nano Benchmark**:
   - Benchmarks on the Jetson Orin Nano Super 8GB would be fantastic! That would provide valuable real-world data on edge embedded deployment for SmolVLA.

**Comment by sattyamjjain** [comment]
Confirmed your point 1 from the checkpoint itself, so nobody has to take it on trust. lerobot/smolvla_base config.json has pad_language_to: max_length and tokenizer_max_length: 48, with num_steps 10 and use_cache true. Prefix length is static at 48 by default.

On the Jetson run, the plan is one thing I will check before any timing.

The thing most likely to invalidate the comparison is whether Triton is present and working on the JetPack 6.x aarch64 torch build. max-autotune needs it to benchmark candidate kernels. If it is missing or stubbed, the compile path either fails or quietly gives something that is not what you measured on the 4090, and the numbers would not be comparable. I will check that first and report it either way, since the answer is useful on its own for anyone deploying this to Orin.

Assuming it works: eager vs compile_denoise on the default 48-token prefix, plus peak memory for both. Orin Nano Super 8GB shares memory between CPU and GPU, so whether it fits is as load-bearing as how fast it runs, and neither number exists anywhere yet.

The number I most want in this thread is warm-up. Yours is 7 to 8 seconds on a 4090. Autotune codegen is CPU-bound and Orin Nano has six A78AE cores, so I would expect minutes rather than seconds. That is my original question in a harsher form: a multi-minute first-call stall is a deployment problem on edge in a way that 8 seconds is not, and it would argue for a warm-up call at policy load rather than on first inference.

One framing point on the title, which I think strengthens the claim rather than weakening it. The checkpoint has chunk_size and n_action_steps both at 50, so one inference produces 50 actions. At 30 fps, that is roughly 1.67 seconds of horizon for 207 ms of compute, which makes 4.8 Hz the chunk rate rather than the control rate. The speedup is still worth having, for latency to first action and for headroom to replan more often. But "limits inference to 4.8 Hz" reads as though the control loop runs at 4.8 Hz, and on an edge box that distinction is the difference between a nice-to-have and a blocker. Being precise about it should help the PR through review.
